src/main/java/net/neoforged/neoforge/items/IItemHandler.java

```diff
@@ -99,4 +103,29 @@ public interface IItemHandler {
      *         false if the slot can never insert the ItemStack in any situation.
      */
     boolean isItemValid(int slot, ItemStack stack);
+
+    @Override
+    default Iterator<ItemStack> iterator() {
+        return new Iterator<>() {
+            private int index = 0;
+
+            @Override
+            public boolean hasNext() {
+                return index < getSlots();
```

Comment:
The `getSlots()` should be cached in a field on creating the iterator, a typical implementation of the `IItemHandler` should not support adding/removing slots, especially in iteration.
The `next()` and `forEachRemaining()` method should also make use of the field to have a better performance.

Reply:
"should not" doesn't mean "will not", especially since this Iterator has no ConcurrentModification protection. It isn't unreasonable to assume some mod will dynamically scale the slots they expose to, e.g, always one per item type plus exactly one empty one.

And if the handler has a static number of slots, getSlots() will be a standard getter and return the value of a field, which is basically as performant as copying that to a field.

Reply:
> "should not" doesn't mean "will not", especially since this Iterator has no ConcurrentModification protection. It isn't unreasonable to assume some mod will dynamically scale the slots they expose to, e.g, always one per item type plus exactly one empty one.

These specific implementation of `IItemHandler` should then provide their own implementation of `iterator()`, there's no reason for a `default` method trying to support every implementation, especially a `IItemHandler` that exposes non-continuous slot numbers is already unsupported by this iterator.